### ISSUE TITLE
TopicUnauthorized error keeps on popping and not the kafka records

### ISSUE THREAD

**Issue by dshivashankar1994** [edenhill/librdkafka#3072]
Description
===========
Coming from - https://gitter.im/edenhill/librdkafka?at=5f58b205a5788a3c29e3076b

Consumer, while subscribing to an unauthorized topic, keeps on emitting the TopicUnauthorizedException and doesn't even proceed to subscribe.

As per Magnus Edenhill, the code currently checks for authorization failure when looking up the topic, which it does prior to subscribe()ing. And in this case it seems like it does have Describe access to the topic, but not Read.

**I see the problem, if a client subscribes to a non-existing topic and later the topic gets created without permissions, the same behavior is observed**

How to reproduce
================
This is easily reproducible with the below change to rdkafka_complex_consumer_example.cpp

```
index 452247db..8e00f67e 100644
--- a/examples/rdkafka_complex_consumer_example.cpp
+++ b/../librdkafka-1.5.0/examples/rdkafka_complex_consumer_example.cpp
@@ -167,9 +167,7 @@ void msg_consume(RdKafka::Message* message, void* opaque) {
   switch (message->err()) {
     case RdKafka::ERR__TIMED_OUT:
       break;
+   case RdKafka::ERR_TOPIC_AUTHORIZATION_FAILED:
+      std::cout << "Unauthorized topic" << std::endl;
+      break;
     case RdKafka::ERR_NO_ERROR:
       /* Real message */
       msg_cnt++;
```

Output:
```
rdkafka_complex_consumer_example_cpp -b mwkafka-staging-01.nyc.firm.com:9092,mwkafka-staging-02.nyc.firm.com:9092,mwkafka-staging-01.dr.firm.com:9092,mwkafka-staging-02.dr.firm.com:9092,mwkafka-staging-01.tbd.firm.com:909 -g new_cons_gp -X security.protocol=SASL_PLAINTEXT -X sasl.kerberos.service.name=kafka -X sasl.mechanism=GSSAPI -X sasl.kerberos.kinit.cmd='skinit --force' -X client.id=cppcon  fps.dhamodha.test-n-9 fps.dhamodha.test-n-10 
ucastest acquired ticket: client ucastest/hostbased@UNIX.firm.COM, server krbtgt/UNIX.firm.COM@UNIX.firm.COM
% Created consumer cppcon#consumer-1
2020-09-09 07:00:13.607: LOG-3-FAIL: [thrd:sasl_plaintext://mwkafka-staging-01.tbd.firm.com:909/bootstra]: sasl_plaintext://mwkafka-staging-01.tbd.firm.com:909/bootstrap: Connect to ipv4#10.236.30.7:909 failed: Connection refused (after 269ms in state CONNECT)
2020-09-09 07:00:13.607: ERROR (Local: Broker transport failure): sasl_plaintext://mwkafka-staging-01.tbd.firm.com:909/bootstrap: Connect to ipv4#10.236.30.7:909 failed: Connection refused (after 269ms in state CONNECT)
RebalanceCb: Local: Assign partitions: fps.dhamodha.test-n-10[0], fps.dhamodha.test-n-9[0], 
Unauthorized topic
Unauthorized topic
Unauthorized topic
Unauthorized topic
Unauthorized topic
Unauthorized topic
Unauthorized topic
```

Log file at : https://gitlab.com/dshivashankar1994/librdkafka-topic-error/-/blob/master/error%20file

Checklist
=========

**IMPORTANT**: We will close issues where the checklist has not been completed.

Please provide the following information:

 - [x] librdkafka version - **Latest**
 - [ ] librdkafka client configuration: `<REPLACE with e.g., message.timeout.ms=123, auto.reset.offset=earliest, ..>`
 - [ ] Operating system: `<REPLACE with e.g., Centos 5 (x64)>`
 - [ ] Provide logs (with `debug=..` as necessary) from librdkafka
 - [ ] Provide broker log excerpts
 - [ ] Critical issue

**Comment by edenhill** [comment]
The Java consumer does the same, by raising an exception to the application and if the application calls poll() again the Fetch will be retried.
It is up to the application to handle this case, either by calling subscribe() again with a new subscription without the errored topic, or by shutting down, or whatever action it deems it needs to take.

What we should do though is increase the fetch error backoff for these types of errors to maybe 10-30s.

**Comment by dshivashankar1994** [comment]
I believe TopicUnauthorized and Non-existing topics should behave the same way. For non-existing topics, we go ahead with subscription and once the topic is created the topic is assigned after the metadata refresh.

Similary for if the permissions are not present, the error should come once and should automatically subscribed to the rest of the topics. Once the topic's permissions issues are resolved, the topic should be picked automatically with the topic.metadatarefresh. 

This seems more natural rather than having different behavior for non-existing and non-authorized topics. 

Could you relook at this ?

**Comment by dshivashankar1994** [comment]
Also, what is the expected behavior in the below case:

Behavior I observe:
1. Consumer subscribing to non-existing topic
2. An error message with the unknown topic is received and the consumer proceeds with the subscription
3. The topic is created but the consumer doesn't have read permissions on this topic
4. The behavior same as above is observed - that is, the unauthorized exception keeps on coming and the records are never received.

Is this the expected behavior ?

**Comment by edenhill** [comment]
If ACLs permit Describe (Metadata requests) but deny Read (Fetch) the consumer will see the topic in all its glory in Metadata response, and will start its fetchers, then get a Auth error in the Fetch response where it can mark the topic as errored, but the next metadata request will see the topic as existent again, and the saga starts over.

We could back off the next fetch a longer time than fetch.error.backoff.ms, but it would eventually pop up again, so I still believe this is the responsibility of the application to decide what to do.


As for your latest question; the consumer will omit the non-existent topics when joining the group and thus not try to consume them.

**Comment by edenhill** [comment]
What we could do is suppress identical fetch errors per partition so it would only be raised once per partition.
The last error would be cleared on assign() and seek().

**Comment by dshivashankar1994** [comment]
>> As for your latest question; the consumer will omit the non-existent topics when joining the group and thus not try to consume them.

I didn't get this part. Is this answer to - https://github.com/edenhill/librdkafka/issues/3072#issuecomment-693180543 ?

If yes, the question is not regarding the non-existing topics completely. What happens if the non-existing topic get created without read permissions. Currently, I see that i keep on getting the error messages and not the topics at all. Are both the problems same ?

**Comment by edenhill** [comment]
With the following PR we change the behaviour when TOPIC_AUTHORIZATION_FAILED is encountered when Fetching:
 - Raise the error only once per partition to the application (through consumer errors)
 - Backoff the next Fetch for the partition `fetch.error.backoff.ms` * 10.

The application is still responsible for making a decision on what to do.

**Comment by edenhill** [comment]
https://github.com/edenhill/librdkafka/pull/3074

**Comment by dshivashankar1994** [comment]
Just to be clear, even after the fix,  the API will not fetch the Kafka records if there is Authorization error. Am i right ?

**Comment by edenhill** [comment]
It will backoff the next fetch for `fetch.error.backoff.ms` * 10, and then try again, but if it hits the error again it will not raise it to the application.

**Comment by dshivashankar1994** [comment]
If there is some other topics in the subscription list, will the records on those topics are consumed ? (if there is an unauthorized topic in the subscription list)

**Comment by edenhill** [comment]
Yes, remaining topics will continue to be fetched.

**Comment by dshivashankar1994** [comment]
When is the commit expected to be released ?

**Comment by edenhill** [comment]
It was released in v1.5.3, but v1.6.0 is now available.
